src/Microsoft.Extensions.Caching.Memory/MemoryCache.cs

```diff
@@ -280,14 +281,36 @@ private static void ScanForExpiredItems(MemoryCache cache)
             }
         }
 
+        private void TriggerOvercapacityCompaction()
+        {
+            if (!_compactionSemaphore.Wait(0))
+            {
+                // Another compaction is running, exit immediately.
+                // Avoid overpurging when multiple overcapacity compactions are triggered concurrently.
+                return;
+            }
+
+            try
+            {
+                // Keep compacting until remaining entries are at 90% of maximum
+                // Stop compacting if no entries were removed, for example if all the remaining entries are pinned with NeverRemove priority
+                while (Compact(1 - ((0.9 * _entryCountLimit.Value) / _entries.Count)) && _entries.Count > _entryCountLimit.Value) { }
```

Comment:
This was triggered by already reaching the max, so why not just ask to compact 10%?

No need for the loop, if it fills up again it will trigger compaction again.